Make macro characters configurable

pyrmont opened the pull request:
> [!NOTE]
> This post has been updated since the original creation of the PR. The text below is the updated post and describes the current situation after subsequent commits were made.

vim-sexp currently hard codes the macro characters that it understands. This means that if a user-defined Lisp uses different macro characters to the default Lisp (i.e. Scheme), then features of vim-sexp that depend on macro characters will not work correctly.

An alternative approach is to allow a user to configure the macro characters, similar to how a user can configure supported filetypes. To achieve this, a global variable `g:sexp_filetype_macro_characters` is checked first, then the script-local variable `s:filetype_macro_characters` (renamed to be consistent with the global variable) is checked, before falling back to the `s:default_macro_characters` if `&lisp` is set (current behaviour).

pyrmont committed:
Make macro characters configurable

vim-sexp currently hard codes the macro characters that it understands.
This means that if a user-defined Lisp uses different macro characters
to the default Lisp (i.e. Scheme), then features of vim-sexp that depend
on macro characters will not work correctly.

An alternative approach is to allow a user to configure the macro
characters, similar to how a user can configure supported filetypes. To
achieve this, the script-local variable `s:macro_filetype_characters` is
changed to the global variable `g:sexp_filetype_macro_characters`.  The
name was selected to be consistent with `g:sexp_filetypes`.

bpstahlman commented on doc/vim-sexp.txt:
```diff
@@ -973,6 +973,20 @@ wish to activate vim-sexp mappings.
+supported |FileTypes|.
+>
+        " Default
+        let g:sexp_filetype_macro_characters = {
```
I think it would be better if the user could override the keys individually. Forcing him to copy the entire dict just to modify one filetype feels like unnecessary redundancy, which would have the unintended consequence of rendering the user's configuration immune to changes to the plugin's defaults. By contrast, if the override were at the filetype level, the user would still benefit from upstream changes involving filetypes he has not explicitly overridden. If `g:sexp_filetype_macro_characters` contained only overridden keys, something like the following could be used to determine the filetype-specific value:
```
get(g:sexp_filetype_macro_characters, &filetype,
    \ get(s:sexp_filetype_macro_characters, &filetype, s:default_macro_characters))
```

pyrmont replied:
Thanks for the feedback. I’ve added a new commit that tries to do this (while retaining existing behaviour of only providing default characters if `&lisp` is true).

bpstahlman replied:
No problem. Thanks for making that change.
I plan to merge this PR but probably not till after I've reached a stopping point on my own branch.

pyrmont replied:
Sounds good! Thanks!

pyrmont committed:
Retain script-local variable but override with global variable if set

pyrmont committed:
Fix missing return statement

bpstahlman reviewed (changes requested):
Found issue during pre-merge testing: specifically, error thrown when user hasn't overridden the global option. Pushed a fix and submitted a PR on your fork.

bpstahlman committed:
Fix error caused by nonexistent g:sexp_filetype_macro_characters

Also, updated docs to advise against copying default array to override a
subset of filetypes.

pyrmont commented:
@bpstahlman Thanks and apologies for the delay. I didn’t get a notification of the PR but did for the message you just sent here. Weird. I’ve merged that change in! Thanks!

bpstahlman commented:
@pyrmont No problem. I probably should have tagged you. I'll most likely merge this next time I get to a temporary stopping place on my actively developed feature branch.
